compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerCoalescedDMAToGlobalLoads.cpp

```diff
@@ -0,0 +1,292 @@
+      return failure();
+    }
+    LDBG() << "Memory layout verified";
+
+    Value source = dmaOp.getSource();
+    Value dest = dmaOp.getInit();
+    auto indicesRange = dmaOp.getIndices();
+
+    auto sourceType = cast<MemRefType>(source.getType());
+    auto destType = cast<MemRefType>(dest.getType());
+
+    if (!indicesRange.empty()) {
+      Value indices = indicesRange.front();
+      LDBG() << "  Indices: " << indices;
+      auto indicesType = cast<MemRefType>(indices.getType());
+      ArrayRef<int64_t> indicesShape = indicesType.getShape();
+      ArrayRef<int64_t> destShape = destType.getShape();
+
+      // Verify that indices dimensions are a prefix of dest dimensions.
+      if (indicesShape.size() > destShape.size()) {
+        return rewriter.notifyMatchFailure(dmaOp,
+                                           "indices rank exceeds dest rank");
+      }
+
+      for (size_t i = 0; i < indicesShape.size(); ++i) {
+        if (indicesShape[i] != destShape[i]) {
+          return rewriter.notifyMatchFailure(
+              dmaOp, "indices shape is not a prefix of dest shape");
+        }
+      }
+    }
+
+    Type elementType = sourceType.getElementType();
+    int64_t elementBits = sourceType.getElementTypeBitWidth();
+    LDBG() << "  Element type: " << elementType;
+    LDBG() << "  Element bits: " << elementBits;
+
+    int64_t innermostDimSize = sourceType.getShape().back();
+    int64_t transferSizeBits = innermostDimSize * elementBits;
+    LDBG() << "  Source innermost dimension size: " << innermostDimSize;
+    LDBG() << "  Transfer size in bits: " << transferSizeBits;
+
+    std::optional<int64_t> subgroupSize =
+        getSubgroupSize(dmaOp->getParentOfType<FunctionOpInterface>());
+    if (!subgroupSize.has_value()) {
+      LDBG() << "FAILED - unable to determine subgroup size";
+      return rewriter.notifyMatchFailure(
+          dmaOp, "unable to determine subgroup size from forall");
+    }
+    LDBG() << "Subgroup size: " << *subgroupSize;
+
+    // Check that transfer size matches one of the target DMA sizes.
+    int64_t transferSizePerLane = transferSizeBits / *subgroupSize;
+    LDBG() << "Transfer size per lane: " << transferSizePerLane << " bits";
+
+    if (!targetDmaSizes.empty() &&
+        !matchesTargetDmaSize(transferSizePerLane, targetDmaSizes)) {
+      LDBG() << "FAILED - transfer size " << transferSizePerLane
+             << " does not match any target DMA size";
+      return rewriter.notifyMatchFailure(
+          dmaOp, "transfer size does not match any target DMA size");
+    }
+    LDBG() << "Transfer size matches target DMA sizes";
+
+    // TODO: Handle indices properly - for now we skip the explicit indices
+    // check The lane parameter is always present but we handle it differently.
+
+    ArrayRef<int64_t> sourceShape = sourceType.getShape();
+    LDBG() << "Source rank: " << sourceShape.size();
+    if (sourceShape.size() < 2) {
```

Comment:
... Why?

Reply:
Matmul we focus on 2D,  but yeah added a TODO here to make sure we handle more cases.